src/allocator_defrag.c

```diff
@@ -0,0 +1,476 @@
+static jeBinsConf arena_bin_conf = {0, NULL, {{0}, 0}, {{0}, 0}};
+static jeUsageLatest usage_latest = {NULL, {0}};
+
+
+/* -----------------------------------------------------------------------------
+ * Alloc/Free API that are cooperative with defrag
+ * -------------------------------------------------------------------------- */
+/* Allocation and free functions that bypass the thread cache
+ * and go straight to the allocator arena bins.
+ * Currently implemented only for jemalloc. Used for online defragmentation. */
+void *allocatorDefragAlloc(size_t size) {
+    void *ptr = je_mallocx(size, MALLOCX_TCACHE_NONE);
+    return ptr;
+}
+void allocatorDefragFree(void *ptr, size_t size) {
+    if (ptr == NULL) return;
+    je_sdallocx(ptr, size, MALLOCX_TCACHE_NONE);
+}
+
+/* -----------------------------------------------------------------------------
+ * Helper functions for jemalloc translation between size and index
+ * -------------------------------------------------------------------------- */
+#define LG_QUANTOM_8_FIRST_POW2 3
+#define SIZE_CLASS_GROUP_SZ 4
+
+#define LG_QUANTOM_OFFSET_3 ((64 >> LG_QUANTOM_8_FIRST_POW2) - 1)
+#define LG_QUANTOM_OFFSET_4 (64 >> 4)
+
+#define getBinindNormal(_sz, _offset, _last_sz_pow2)                                                             \
+    ((SIZE_CLASS_GROUP_SZ - (((1 << (_last_sz_pow2)) - (_sz)) >> ((_last_sz_pow2) - LG_QUANTOM_8_FIRST_POW2))) + \
+     (((_last_sz_pow2) - (LG_QUANTOM_8_FIRST_POW2 + 3)) - 1) * SIZE_CLASS_GROUP_SZ + (_offset))
+/* Get the bin index in bin array from the reg_size.
+ *
+ * these are reverse engineered mapping of reg_size -> binind. We need this information because the utilization query
+ * returns the size of the buffer and not the bin index, and we need the bin index to access it's usage information
+ *
+ * Note: In case future PR will return the binind (that is better API anyway) we can get rid of
+ * these conversion functions*/
+inline unsigned jeSize2BinIndexLgQ3(size_t sz) {
+    if (sz <= (1 << (LG_QUANTOM_8_FIRST_POW2 + 3))) {
+        // for sizes: 8, 16, 24, 32, 40, 48, 56, 64
+        return (sz >> 3) - 1;
+    }
+    // following groups have SIZE_CLASS_GROUP_SZ size-class that are
+    uint64_t last_sz_in_group_pow2 = 64 - __builtin_clzll(sz - 1);
+    return getBinindNormal(sz, LG_QUANTOM_OFFSET_3, last_sz_in_group_pow2);
+}
+
+/* -----------------------------------------------------------------------------
+ * Get INFO string about the defrag.
+ * -------------------------------------------------------------------------- */
+/*
+ * add defrag info string into info
+ */
+sds allocatorDefragCatFragmentationInfo(sds info) {
+    if (!defrag_supported) return info;
+    jeBinInfo *bin_info;
+    jeBusage *bin_usage;
+    unsigned nbins = arena_bin_conf.nbins;
+    if (nbins > 0) {
+        info = sdscatprintf(info,
+                            "jemalloc_quantum:%d\r\n"
+                            "defrag_hit_ratio:%.2f\r\n"
```

Comment:
We generally avoid ratios and other easily derivable values from info. This number will likely be not useful if the server has been running for a long time as it might have been skewed in the past.

Reply:
removed all togather